subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -225,33 +227,26 @@ private DependencyState maybeSubstitute(DependencyState dependencyState) {
         return dependencyState;
     }
 
-    private List<EdgeState> findTransitiveIncomingEdges(boolean hasIncomingEdges) {
-        if (!hasIncomingEdges) {
-            return Collections.emptyList();
-        }
-
-        int size = incomingEdges.size();
-        if (size == 1) {
-            return findSingleIncomingEdge();
+    /**
+     * Returns the set of incoming edges that are transitive. Most edges are transitive, so the implementation is optimized for this case.
+     */
+    private List<EdgeState> getTransitiveIncomingEdges() {
+        if (isRoot()) {
+            return incomingEdges;
         }
-
-        List<EdgeState> transitiveIncoming = Lists.newArrayListWithCapacity(size);
-        for (EdgeState edge : incomingEdges) {
-            if (edge.isTransitive()) {
-                transitiveIncoming.add(edge);
+        for (EdgeState incomingEdge : incomingEdges) {
+            if (!incomingEdge.isTransitive()) {
+                // Have a non-transitive edge: return a filtered list
+                return CollectionUtils.filter(incomingEdges, new Spec<EdgeState>() {
```

Comment:
I guess it would be worth extracting the spec as a constant.

Reply:
Roger that.